internal/security/setup/option/cache_test.go

```diff
@@ -0,0 +1,209 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+//
+// SPDX-License-Identifier: Apache-2.0'
+//
+package option
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/edgexfoundry/edgex-go/internal/security/setup"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCacheOn(t *testing.T) {
+	vaultJSONPkiSetupExist = true
+	kongJSONPkiSetupExist = true
+	tearDown := setupCacheTest(t)
+	defer tearDown(t)
+
+	options := PkiInitOption{
+		CacheOpt: true,
+	}
+	cacheOn, _, _ := NewPkiInitOption(options)
+	cacheOn.(*PkiInitOption).executor = testExecutor
+
+	var exitStatus exitCode
+	var err error
+	f := Cache()
+	exitStatus, err = f(cacheOn.(*PkiInitOption))
+
+	cacheEmpty, emptyErr := isDirEmpty(getPkiCacheDirEnv())
+
+	assert := assert.New(t)
+	assert.Equal(normal, exitStatus)
+	assert.Nil(err)
+	assert.Nil(emptyErr)
+	assert.False(cacheEmpty)
+
+	generatedDirPath := filepath.Join(os.Getenv(envXdgRuntimeDir), pkiInitGeneratedDir)
+	caPrivateKeyFile := filepath.Join(generatedDirPath, caServiceName, tlsSecretFileName)
+	if _, checkErr := os.Stat(caPrivateKeyFile); checkErr == nil {
+		// means found the file caPrivateKeyFile
+		assert.Fail("CA private key are not removed!")
+	}
+
+	deployEmpty, emptyErr := isDirEmpty(pkiInitDeployDir)
+	assert.Nil(emptyErr)
+	assert.False(deployEmpty)
+}
+
+func TestCacheDirNotEmpty(t *testing.T) {
+	vaultJSONPkiSetupExist = true
+	kongJSONPkiSetupExist = true
+	tearDown := setupCacheTest(t)
+	defer tearDown(t)
+
+	options := PkiInitOption{
+		CacheOpt: true,
+	}
+	cacheOn, _, _ := NewPkiInitOption(options)
+	cacheOn.(*PkiInitOption).executor = testExecutor
+
+	var exitStatus exitCode
+	var err error
+	f := Cache()
+	exitStatus, err = f(cacheOn.(*PkiInitOption))
+
+	cacheEmpty, emptyErr := isDirEmpty(getPkiCacheDirEnv())
+
+	assert := assert.New(t)
+	assert.Equal(normal, exitStatus)
+	assert.Nil(err)
+	assert.Nil(emptyErr)
+	assert.False(cacheEmpty)
+
+	generatedDirPath := filepath.Join(os.Getenv(envXdgRuntimeDir), pkiInitGeneratedDir)
+	// now we move the whole generated directory and leave the cache dir untouched
+	os.RemoveAll(generatedDirPath)
+
+	// call the cache option again:
+	exitStatus, err = f(cacheOn.(*PkiInitOption))
+
+	cacheEmpty, emptyErr = isDirEmpty(getPkiCacheDirEnv())
+
+	assert.Equal(normal, exitStatus)
+	assert.Nil(err)
+	assert.Nil(emptyErr)
+	assert.False(cacheEmpty)
+
+	deployEmpty, emptyErr := isDirEmpty(pkiInitDeployDir)
+	assert.Nil(emptyErr)
+	assert.False(deployEmpty)
+}
+
+func TestCacheOff(t *testing.T) {
+	vaultJSONPkiSetupExist = true
+	kongJSONPkiSetupExist = true
+	tearDown := setupCacheTest(t)
```

Comment:
is there a reason not to just do 
```suggestion
	defer setupCacheTest(t)()
```

or if you do really need to be able to pass in `t` to the function returned:

```suggestion
	defer setupCacheTest(t)(t)
```

Reply:
I think it is more readable if we call that variable name out, ie. `tearDown` instead of inline coding.

Reply:
okay actually I think you're right and defining it is better, I take back my suggestion